### PR TITLE
feat: .NET Core Support

### PR DESCRIPTION
**Pull request by lokitoth**
WIP branch for .NET Core Support

Needs:

- [x] Rebase to VW current
- [x] Merge in: Linux build
- [x] Merge in: NuGet generation

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -6,5 +6,55 @@
             ],
```
Comment (lokitoth) [inline .vscode/settings.json:6]: - [ ] delete this

File: cs/CMakeLists.txt.old [hunk 2]
```diff
@@ -0,0 +1,32 @@
+message(STATUS "Building .NET Framework Bindings")
```
Comment (lokitoth) [inline cs/CMakeLists.txt.old:1]: - [ ] delete this

File: cs/unittest/Test1and2.cs [hunk 3]
```diff
@@ -45,6 +45,9 @@ public void Test1and2()
 
                         Assert.AreEqual((float)expected, actual, 1e-6, "Learn output differs on line: " + lineNr);
 
+                        ulong vwStr_examplesPerPass = vwStr.PerformanceStatistics.NumberOfExamplesPerPass;
```
Comment (lokitoth) [inline cs/unittest/Test1and2.cs:48]: - [ ] debug spew leftovers

**Comment by rajan-chari** [comment]
Nice!

File: cmake/platforms/win32.cmake [hunk 4]
```diff
@@ -19,6 +22,26 @@ if(WIN32)
...
+    # windows and redirect all targets to the same place.
+    SET(vw_win32_CMAKE_RUNTIME_OUTPUT_DIRECTORY_backup ${CMAKE_RUNTIME_OUTPUT_DIRECTORY})
+
+    if (vw_DOTNET_USE_MSPROJECT)
+      SET(CMAKE_RUNTIME_OUTPUT_DIRECTORY "${CMAKE_BINARY_DIR}/binaries/x64")
+    else()
+      SET(CMAKE_RUNTIME_OUTPUT_DIRECTORY "${CMAKE_BINARY_DIR}/binaries/")
+    endif()
+
+    add_custom_target(launch_vs COMMAND set "BinaryOutputBase=${CMAKE_BINARY_DIR}/binaries" && start "${CMAKE_BINARY_DIR}/vowpal_wabbit.sln")
```
Comment (rajan-chari) [inline cmake/platforms/win32.cmake:42]: This is awesome.

File: cs/common/Reflection/ReflectionHelper.cs [hunk 5]
```diff
@@ -31,6 +31,10 @@ public static class ReflectionHelper
         /// <remarks>Can't constraint on Func (or would have to have 11 overloads) nor is it possible to constaint on delegate.</remarks>
         public static System.Delegate CompileToFunc<T>(this Expression<T> sourceExpression)
         {
+#if NETSTANDARD
```
Comment (rajan-chari) [inline cs/common/Reflection/ReflectionHelper.cs:34]: Check for mis-merge.

File: cs/unittest/TestAllReduce.cs [hunk 6]
```diff
@@ -44,6 +44,7 @@ private static void Ingest(VowpalWabbitAsync<CbAdfShared, CbAdfAction> vw, IEnum
 
         [TestMethod]
         [TestCategory("Vowpal Wabbit")]
+        //[Ignore]
```
Comment (rajan-chari) [inline cs/unittest/TestAllReduce.cs:47]: remove line

File: cs/unittest/TestLeak.cs [hunk 7]
```diff
@@ -2,20 +2,20 @@
```
Comment (rajan-chari) [inline cs/unittest/TestLeak.cs:2]: Delete file

**Review by rajan-chari — APPROVED** [review]
Looks good to me.  Let's ship!
⛵

**Comment by PeterDraex** [comment]
I'm excited for this :)

**Comment by PeterDraex** [comment]
When will this get released? :)

**Comment by olgavrou** [comment]
@PeterDraex probably first couple weeks of September

**Comment by PeterDraex** [comment]
> @olgavrou: probably first couple weeks of September

Hi, is there any update on the expected release date?

**Pull request merged**